include/avif/avif.h

```diff
@@ -7,6 +7,20 @@
 #include <stddef.h>
 #include <stdint.h>
 
+#ifndef AVIF_API
```

Comment:
If the user has enough knowledge on this, then the user can define `AVIF_API` him/herself, and we just don't mess up user's decision. Also dav1d and libgav1 did this check too.

Reply:
Thanks for the reply. I didn't realize the #ifndef AVIF_API test is intentional. Is this intended to allow a user to add -DAVIF_API="__declspec(dllimport)" to the compiler command line if the user knows the libavif library is a DLL?

We should document why a user may want to define AVIF_API. (I am very familiar with build systems, but I could not figure this out.) I personally would prefer that we remove the #ifndef AVIF_API test because allowing a user to define AVIF_API seems error-prone and unlikely to be useful.

Note: I didnt know dav1d and libgav1 also do this check. (I actually reviewed libgav1's symbol_visibility.h header when it was added but missed this.) While investigating this issue, I noticed that dav1d and libgav1 both use the "default" visibility attribute on their public API functions when they are built as static libraries. This has the undesirable effect that the dav1d and libgav1 public API functions used by libavif are also exported from libavif.so if we link libavif.so with the static dav1d and libgav1 libraries.

Reply:
My initial thought was not to force user to edit headers of dependencies, so a user can at least do this:
```c
#define AVIF_API __declspec(dllimport)
#include <avif/avif.h>
```
If preventing the overhead is really a need. But surely having another macro named something like `AVIF_DLL` is better. 

Note: I'm more concerned about libaom as it does nothing to control visibility, so libavif.so even exports non-API functions from libaom, and make the export table bloating (there's 4000+ private functions in libaom being exported). This will slow down library loading speed and prevent compiler from inlining private functions to get better performance.

Reply:
Why are you trying to statically link libaom into libavif? If you're going the DLL route, libaom should also be its own DLL, no?

Reply:
Joe: I agree it may not make sense to statically link libaom into libavif. However, it may make sense to statically link libaom into some other shared library, and that use case has the same problem.

Yuan Tong: I also saw hundreds of exported libaom symbols when I built libavif.so with the libaom static library. I believe that problem can be solved by improving libaom's build system. libaom uses symbol files rather than `__declspec(dllexport)` and _`_attribute__ ((visibility ("default")))` to control exported symbols. The relevant files in libaom are:
- build/cmake/exports.cmake
- aom/exports_com, aom/exports_dec, aom/exports_enc, aom/exports_test.

I also looked into the #if !defined(LIBGAV1_PUBLIC) check in libgav1. It turns out to be a feature used by Chrome to work around problem of the exported symbols from the libgav1 static library. I will try to fix that when I have spare time.

I surveyed several open-source projects. I found that many of them have the same problem as dav1d and libgav1. The solution I like the best is the one described in the "Step by step guide" in https://gcc.gnu.org/wiki/Visibility. So I suggest we implement a variant of that solution.

Here are two examples of a variant of that solution. Note that they differ from the example in https://gcc.gnu.org/wiki/Visibility in the definition of the macro in the Unix shared library import case. The example in https://gcc.gnu.org/wiki/Visibility defines the macro as `__attribute__ ((visibility ("default")))` whereas the following two examples defines the macro as empty.

1. BoringSSL: boringssl/src/include/openssl/base.h

```
// BORINGSSL_API_VERSION is a positive integer that increments as BoringSSL
// changes over time. The value itself is not meaningful. It will be incremented
// whenever is convenient to coordinate an API change with consumers. This will
// not denote any special point in development.
//
// A consumer may use this symbol in the preprocessor to temporarily build
// against multiple revisions of BoringSSL at the same time. It is not
// recommended to do so for longer than is necessary.
#define BORINGSSL_API_VERSION 12

#if defined(BORINGSSL_SHARED_LIBRARY)

#if defined(OPENSSL_WINDOWS)

#if defined(BORINGSSL_IMPLEMENTATION)
#define OPENSSL_EXPORT __declspec(dllexport)
#else
#define OPENSSL_EXPORT __declspec(dllimport)
#endif

#else  // defined(OPENSSL_WINDOWS)

#if defined(BORINGSSL_IMPLEMENTATION)
#define OPENSSL_EXPORT __attribute__((visibility("default")))
#else
#define OPENSSL_EXPORT
#endif

#endif  // defined(OPENSSL_WINDOWS)

#else  // defined(BORINGSSL_SHARED_LIBRARY)

#define OPENSSL_EXPORT

#endif  // defined(BORINGSSL_SHARED_LIBRARY)
```

2. protobuf: protobuf/src/google/protobuf/port_def.inc

Note: I omitted some code.

```
#if defined(PROTOBUF_USE_DLLS)
#if defined(_MSC_VER)
...
#ifdef LIBPROTOC_EXPORTS
#define PROTOC_EXPORT __declspec(dllexport)
#else
#define PROTOC_EXPORT __declspec(dllimport)
#endif
#else  // defined(_MSC_VER)
...
#ifdef LIBPROTOC_EXPORTS
#define PROTOC_EXPORT __attribute__((visibility("default")))
#else
#define PROTOC_EXPORT
#endif
#endif
#else  // defined(PROTOBUF_USE_DLLS)
...
#define PROTOC_EXPORT
...
#endif
```